### ISSUE TITLE
[Bug] Y8950 is incorrectly handling key-off for percussive tones [sf#282]

### ISSUE THREAD

**Issue by openMSX-import** [openMSX/openMSX#583]
**Reported by sd-snatcher on 2007-11-28 01:26 UTC**
All OPLx (OPL1 to OPL4) only handle key-off for non-percursive tones.

For the Y8950, this behavior is described on it's datasheet, on Fig 3-1 of page 23.

The current implementation of the Y8950 used on OpenMSX is also handling key-off for the percursive tones, which is wrong.

The Moonsound (OPL3/4) core is doing it correctly.

Best regards,
FRS

**Comment by openMSX-import** [comment]
**Commented by m9710797 on 2007-11-28 21:06 UTC**
Logged In: YES 
user_id=356949
Originator: NO

Hi FRS,

Thanks for this report. However I do have a few questions.

I indeed see a difference in the handling of the keyon/off handling of percussive tones between OPL3 and OPL1. Although it's not really clear that this difference explains the thing in this bug report (as you know we didn't write this code ourself, we ported it from MAME).
How did you find out out about this bug? Do you have a testcase that demonstrates the problem? So that I can verify that when I copy (part of) the code, it really fixes the bug.

In openMSX we have two alternative OPLL (FMPAC) cores, one of the differences between these two cores is exactly the same difference as the keyon/off handling between OPL1 and OPL3. So probably it's wrong there as well. Do you also have info on this for the OPLL? Again, a testcase would be very useful.

Thanks again.

Wouter

**Comment by openMSX-import** [comment]
**Commented by sd-snatcher on 2007-12-12 20:36 UTC**
Logged In: YES 
user_id=130840
Originator: YES

Hi Wouter!

Sorry for the long delay to answer. I'm quite busy nowadays.

The test case for the Y8950 is very easy to set up. You just need to use the MSX-Audio BIOS v1.0 and use the command below on BASIC of both the real MSX and openMSX:

CALL AUDIO
PLAY#2,"@33V15A8"

You'll see that on the real MSX the note will take a long time to go off and on the openMSX it will be cut off shortly. This has an explanation: The original MSX-Audio BIOS had a bug and didn't handled the key-off properly for percussive tones. It just lowered the KON bit of the register BDh for both percussive and non-percussive tones.

On the real MSX (I tested on a Toshibas HX-MU900) this bug can be clearly noticed, but on openMSX it is masqueraded by the incorrect KEY-OFF handling. Five years ago I did also test this KEY-OFF handling on the OPL3 and it behaves just like the OPL1. Please note that I have knowledge of what I'm posting here. I researched this a lot when I developed the YM2413-&gt;YM3812 translator that MAME used for quite some time and I used a real OPL3 for that. Just google for "schmidlin MAME" and you'll see it.

If you don't have the bulky Panasonic FS-CA1 at hand to test it using the AUDIO BASIC, this is not a problem anymore: just install the BIOS on a Philips Music Module or on a Toshiba HX-MU900.  I wrote both articles below. Any doubts about it or problems, just ask.   :)

MSX-Audio BIOS upgrade for the Philips Music Module
http://www.msxpro.com/mm\_upgrade.html
(note: There's an link for the English version on the beginning of the article)

DIY upgrade guide for the Toshiba HX-MU900 (256KB SampleRAM + BIOS)
http://www.hansotten.com/news/toshibahx900upgrades.zip

&gt; So probably it's wrong there as
well. Do you also have info on this for the OPLL? 

Yes, I do have. The YM2413 is the only OPL soundchip that handle KEY-OFFs for the percursive tones. This is described on the YM2413 datasheet on both the graphics of the page-10 and on the SUS-ON/OFF description of the page-14. Yamaha had no choice but do this KEY-OFF handling on the YM2413, since on its simplified register-set you do not have access to the SL/RR register of the channels (the instruments are stored on its internal ROM). 
All other OPL(1-4) soundchips require the KEY-OFF handling for percussive tones to be done by the software. But please note that this is much better than the hardcoded method of the YM2413, as it allows the programmer to adjust it for the specific instrument being played.

In short words, opposed to the OPL1-4 behavior, a YM2413 emulated core must handle KEY-OFFs for percussive-tones.

Best regards and thank you for the excellent openMSX emulator,
FRS

**Comment by openMSX-import** [comment]
**Updated by manuelbi on 2007-12-17 09:57 UTC**
- **summary**: Y8950 is incorrectly handling key-off for percursive tones --> Y8950 is incorrectly handling key-off for percussive tones

**Comment by openMSX-import** [comment]
**Commented by m9710797 on 2014-08-27 20:19 UTC**
Hi FRS,

It's a very long time since you submitted this bug report. I finally fixed it. Or at least attempted to fix it .. I fixed _a_ bug in the percussive key-on/off handling (and now it's the same as how it's emulated for OPL3), I verified my fix on real hardware. But the scenario you described in this bug report still behaves the same before and after my fix. But it's possible I'm not testing with the correct version of the ROM. Can you verify? See also the commit message for the fix (https://sourceforge.net/p/openmsx/openmsx/ci/903c74a526cdc82170ef5241d117496df8edf415/).

**Comment by openMSX-import** [comment]
**Commented by m9710797 on 2014-08-27 20:57 UTC**
Never mind seems like I confused percussive with rhythm. I'll make another attempt at fixing this soon. (And hey, I did fix another bug ;-)

**Comment by openMSX-import** [comment]
**Commented by m9710797 on 2014-08-28 19:28 UTC**
- Description has changed:

Diff:

```

--- old
+++ new
@@ -1,4 +1,3 @@
-
 All OPLx \(OPL1 to OPL4\) only handle key-off for non-percursive tones.

 For the Y8950, this behavior is described on it's datasheet, on Fig 3-1 of page 23.

```
- **status**: open --> closed-fixed
- **Group**:  --> Next_release

**Comment by openMSX-import** [comment]
**Commented by m9710797 on 2014-08-28 19:28 UTC**
Should be fixed now (please verify).
Revision eeebb748.
